src/buffer/mod.rs

```diff
@@ -1961,83 +1961,343 @@ impl TextBuffer {
+        let final_cursor_pos = self.cursor_move_to_logical_internal(cursor_at_line_start, new_selection_start);
+        self.set_cursor_internal(final_cursor_pos);
+    }
+
+    fn handle_selection_backward_indent(&mut self, original_selection_start: Point, original_selection_end: Point, tab_size: usize) {
+        // store the current cursor for consistent base calculations.
+        let current_cursor = self.cursor;
+
+        // get precise cursor positions for the beginning and end of the selection.
+        let selection_beg_cursor = self.cursor_move_to_logical_internal(current_cursor, original_selection_start);
+        let selection_end_cursor = self.cursor_move_to_logical_internal(current_cursor, original_selection_end);
+
+        // extract the bytes of the currently selected text.
+        let mut selected_bytes = Vec::new();
+        self.buffer.extract_raw(
+            selection_beg_cursor.offset,
+            selection_end_cursor.offset,
+            &mut selected_bytes,
+            0
+        );
 
-            let mut remove = 0;
+        // convert selected bytes to a string for character-based checks.
+        let selected_text_str = std::str::from_utf8(&selected_bytes).unwrap_or("");
 
-            if replacement[offset] == b'\t' {
-                remove = 1;
+        // determine if the *entire* selected text consists only of spaces or tabs.
+        let is_selected_whitespace_only = selected_text_str.chars().all(|c| c == ' ' || c == '\t');
+
+        // handle highlighted text being only whitespace (spaces/tabs)
+        if is_selected_whitespace_only {
+            let delete_start_actual_cursor = selection_beg_cursor;
+            let delete_end_actual_cursor = selection_end_cursor;
+
+            // ensure there's actually a range to delete.
+            if delete_start_actual_cursor.offset < delete_end_actual_cursor.offset {
+                // perform the deletion within an edit transaction.
+                self.edit_begin(HistoryType::Other, delete_start_actual_cursor);
+                self.edit_delete(delete_end_actual_cursor);
+                self.edit_end();
+
+                // after deletion, the selection is consumed, so clear it.
+                self.selection = None;
+                // place the cursor at the point where the selection originally started.
+                self.set_cursor_internal(delete_start_actual_cursor);
             } else {
-                while remove < self.tab_size as usize
-                    && offset + remove < replacement.len()
-                    && replacement[offset + remove] == b' '
-                {
-                    remove += 1;
-                }
+                // log a warning if an attempt was made to delete an empty whitespace selection.
+                eprintln!("Warning: Attempted to delete empty whitespace selection. No edit performed.");
```

Comment:
FYI: We can't print to stderr in this app. When the target is a TTY (a terminal) then stdout and stderr are the exact same pipe. So, this would write this text on top of the current screen contents.